render/preprocessor/preprocessor.go

```diff
@@ -257,82 +257,130 @@ func HasMethod(obj interface{}, methodName string) bool {
 
 // EvalMember gets a member from a stuct, dereferencing pointers as necessary
 func EvalMember(name string, obj interface{}) (reflect.Value, error) {
-	name, err := LookupCanonicalFieldName(interfaceToConcreteType(obj), name)
-
-	if err != nil {
-		return reflect.Zero(reflect.TypeOf(obj)), err
+	keys, fields := lookupMap(FieldMap(obj))
+	k, ok := keys[strings.ToLower(name)]
+	if !ok {
+		var validValues []string
+		for k := range keys {
+			validValues = append(validValues, k)
+		}
+		return reflect.ValueOf(obj), fmt.Errorf("%T has no field %s. Must be one of: %v", obj, name, validValues)
 	}
-	v := reflect.ValueOf(obj)
-	for v.Kind() == reflect.Ptr || v.Kind() == reflect.Interface {
-		if v.IsNil() {
-			return reflect.Zero(reflect.TypeOf(obj)), nilPtrError(v)
+	return fields[k], nil
+}
+
+// FieldMap generates a map of field names to values for an interface, including
+// embedded structs and interfaces.  If a field is defined in more than one
+// embedded struct or interface then it is excluded (unless it is also present
+// in the parent struct).  Fields in obj will take priority over those in
+// embedded structs.
+func FieldMap(obj interface{}) map[string]reflect.Value {
+	embeddedRefs := make(map[string]int)
+	unembeddedRefs := make(map[string]int)
+
+	results := make(map[string]reflect.Value)
+	val := reflect.ValueOf(obj)
+	if canBeNil(val) && val.IsNil() {
+		return results
+	}
+	for ; (val.Kind() == reflect.Ptr && !val.IsNil()) || val.Kind() == reflect.Interface; val = val.Elem() {
```

Comment:
We're mutating val here so that as long as it's a non-nil pointer or an interface we get the inner Elem.  There's nothing to do in the block itself since all the mutation is done in the for.